### PR TITLE
Overhaul Python C layer

### PR DESCRIPTION
**Pull request by soltanmm**
Change the Python C layer to better match the core; currently adapts the new C layer via `_low.py` to what the old `_low.py` looked like (sans a name change `EXPIRED` -> `DEADLINE_EXCEEDED` to match the core enums). This effectively moves the boundary between old-style and new-style APIs from the shared Python/C layer to the lowest pure Python layer.

There is much to be cleaned up in this pull request (file copyrights, doc comments, commit messages, source file organization, enum location [lingual]...), and I'd rather wait for #1472 and #1493 to go through first as they seem like potential sources of surprise (which is also a reason for the alternative, but, meh).

This is being put up as a PR to check against Travis (all tests are passing locally) and for initial comments.

### CODE REVIEW

**Comment by soltanmm** [comment]
... and part of cleaning up the commit history will include setting the proper e-mail address for the commit.

File: src/python/src/grpc/_adapter/_blocking_invocation_inline_service_test.py [hunk 1]
```diff
@@ -43,4 +43,4 @@ class BlockingInvocationInlineServiceTest(
 
 
 if __name__ == '__main__':
-  unittest.main()
+  unittest.main(verbosity=2)
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_blocking_invocation_inline_service_test.py]: I've never before seen "verbosity=2" in a call to "unittest.main()". What does it do?

File: src/python/src/grpc/_adapter/_c/module.c [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+                          "Wrappings of C structures and functions.");
+
+  if (pygrpc_module_add_types(module) < 0) {
+    return;
+  }
+
+  /* GRPC maintains an internal counter of how many times it has been
+     initialized and handles multiple pairs of grpc_init()/grpc_shutdown()
+     invocations accordingly. */
+  grpc_init();
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_c/module.c]: Cool; I like this happening in the .c code and not even being visible in any Python code.

File: src/python/src/grpc/_adapter/_c/module.c [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#include <stdlib.h>
+
+#include <grpc/grpc.h>
+
+#include "grpc/_adapter/_c/types.h"
+
+static PyMethodDef c_methods[] = {
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_c/module.c]: Does this have to be defined or can you just write its value at its one use-site?

File: src/python/src/grpc/_adapter/_c_test.py [hunk 4]
```diff
@@ -1,219 +1,40 @@
...
-  def test_secure_server(self):
-    _c.init()
+  def testCompletionQueueUpDown(self):
+    cq = _c.CompletionQueue()
+    del cq
 
-    server_credentials = _c.ServerCredentials(
-        'root certificate', (('private key', 'certificate chain'),))
+  def testServerUpDown(self):
+    cq = _c.CompletionQueue()
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_c_test.py]: Our style guide disallows single-letter identifier names, and initialisms, acronyms, and abbreviations aren't any better. Spell out "completion_queue", "credentials", "server", "channel", and so forth.

File: tools/run_tests/build_python.sh [hunk 5]
```diff
@@ -38,5 +38,5 @@ rm -rf python2.7_virtual_environment
 virtualenv -p /usr/bin/python2.7 python2.7_virtual_environment
 source python2.7_virtual_environment/bin/activate
 pip install enum34==1.0.4 futures==2.2.0 protobuf==3.0.0-alpha-1
-CFLAGS=-I$root/include LDFLAGS=-L$root/libs/$CONFIG pip install src/python/src
+CFLAGS="-I$root/include" LDFLAGS="-L$root/libs/$CONFIG" pip install src/python/src
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/build_python.sh]: The change on this line smells entirely independent of the whole rest of this commit. May this be made in a separate pull request?

Comment (soltanmm) [inline tools/run_tests/build_python.sh]: Huh. I could've sworn I removed that change...

File: src/python/src/grpc/_adapter/_blocking_invocation_inline_service_test.py (hunk 1 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_blocking_invocation_inline_service_test.py]: Changes the real-time normal test output to be the test names rather than `.`s and `S`s and `E`s.

**Comment by soltanmm** [comment]
It appears that the segfault isn't present when building with CONFIG=dbg but is present when CONFIG=opt. Fun stuff.

**Comment by nathanielmanistaatgoogle** [comment]
What's the reason for having large and few C files rather than small and many? Might I push gently to move in the small-and-many direction?

**Comment by soltanmm** [comment]
I did that so that I could have all the functions be declared `static` and still keep the forward declarations of the Python-side 'classes' in one place without resorting to preprocessor ugliness. There's no particular reason to keep everything `static`, though, so I'll split them up.

File: src/python/src/grpc/_adapter/_c/module.c (hunk 3 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_c/module.c]: It has to be defined _somewhere_ at some point (no type inference in C). Would you prefer it to be inside `init_c`?

Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_c/module.c]: The type point was lost on me. It's fine like this in that case.

**Comment by soltanmm** [comment]
We are now caught up to #1472 and #1493.

File: src/python/src/grpc/_adapter/_c/types.c [hunk 6]
```diff
@@ -1,3 +1,4 @@
+/* vim: set expandtab shiftwidth=2 tabstop=2 softtabstop=2: */
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_c/types.c]: Please drop these lines?

File: src/python/src/grpc/_adapter/rear.py [hunk 7]
```diff
@@ -35,7 +35,7 @@
 import time
 
 from grpc._adapter import _common
-from grpc._adapter import _low
+import grpc._adapter._intermediary_low as _low
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/rear.py]: Why is this not a from-import? It looks strange among its neighbors.

File: src/python/src/grpc/_adapter/_types.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
+# vim: set expandtab shiftwidth=2 tabstop=2 softtabstop=2:
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_types.py]: Drop this line too?

File: src/python/src/grpc/_adapter/_links_test.py [hunk 9]
```diff
@@ -110,6 +110,7 @@ def rear_action(front_to_back_ticket, fore_link):
           test_fore_link.tickets[-1].kind,
           interfaces.BackToFrontTicket.Kind.COMPLETION)
 
+  @unittest.skip('')
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_links_test.py]: I'm not sure that I understand the changes to this file?

File: src/python/src/grpc/_adapter/rear.py (hunk 7 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/rear.py]: Because I'd forgotten that one can do `from * import * as *`. :-/

File: src/python/src/grpc/_adapter/_links_test.py (hunk 9 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_links_test.py]: Forgetfulness. :-(

File: src/python/src/grpc/_adapter/_c/types.c (hunk 6 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_c/types.c]: Werp - I'd forgotten I had those there.

**Comment by soltanmm** [comment]
Added `DO NOT MERGE`; there's a long list of error-handling clean-up tasks for this PR.

**Comment by ctiller** [comment]
Is there a timeline for merging this?
I ask not to try and accelerate it, but simply to pace my own work.

**Comment by soltanmm** [comment]
Not really - what remains is cleaning up error handling and getting reviewed by someone from the Python team. I've kind of put off cleaning this up in favor of getting Python build on Windows to establish whether or not we'd be likely to hit problems if we moved to using the C99 standard. I'm willing to suffer C89's 89ish-ness for now if this is some manner of eventual blocker or source-of-future-wasted-engineering-time-as-long-as-it-isn't-merged.

**Comment by ctiller** [comment]
My reason for asking is #1577, which is not yet urgent. It will require some amount of restructuring to _low.py, which is rewritten as far as I can tell in this change.

Given that we're coming up on a release, I may focus instead on stability for a week or so and leave that change in maintenance mode.

I would encourage sticking to the same C subset as the rest of the project however - it seems better that we're uniform on that than having yet another variation across the project.

**Comment by soltanmm** [comment]
I believe I've covered my error-handling-bases and have dropped @Yhg1s a line - we'll see where this PR ends up in the next few days. I guess, too, that we're sticking with C89 for the foreseeable future.

File: src/python/src/grpc/_adapter/_c/module.c [hunk 10]
```diff
@@ -31,21 +31,31 @@
  *
  */
 
-#ifndef _ADAPTER__SERVER_H_
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_c/module.c]: "src/python/src/grpc/_adapter/_server.h → src/python/src/grpc/_adapter/_c/module.c" is a funny move to make, and makes for a funny diff - how easily could you change this file to being a fork of something more sensible (_c.c?)?

File: src/python/src/grpc/_adapter/_c/types.c [hunk 11]
```diff
@@ -31,19 +31,30 @@
  *
  */
 
-#ifndef _ADAPTER__COMPLETION_QUEUE_H_
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_c/types.c]: "src/python/src/grpc/_adapter/_completion_queue.h → src/python/src/grpc/_adapter/_c/types.c" makes for another funny diff.

File: src/python/src/grpc/_adapter/_types.py [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+      OpType.SEND_INITIAL_METADATA, else is None.
+    trailing_metadata (sequence of 2-sequence of str): Only valid if type ==
+      OpType.SEND_STATUS_FROM_SERVER, else is None.
+    message (bytes): Only valid if type == OpType.SEND_MESSAGE, else is None.
+    status (Status): Only valid if type == OpType.SEND_STATUS_FROM_SERVER, else
+      is None.
+  """
+
+  @staticmethod
+  def send_initial_metadata(init_md):
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_types.py]: Avoid abbreviations unless you have a really good reason. Why shouldn't this method's parameter just be "initial_metadata"?

File: src/python/src/grpc/_adapter/_types.py [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    return OpArgs(OpType.SEND_INITIAL_METADATA, init_md, None, None, None)
+
+  @staticmethod
+  def send_message(msg):
```
Comment (nathanielmanistaatgoogle) [inline src/python/src/grpc/_adapter/_types.py]: "message"?

**Comment by nathanielmanistaatgoogle** [comment]
(Good work, by the way.)

File: src/python/src/grpc/_adapter/_c/module.c (hunk 10 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_c/module.c]: I'll look into `git mv`, but IIRC it's just the equivalent of `git add` + `git rm`.

**Comment by soltanmm** [comment]
I managed to introduce a segfault on local testing.

File: src/python/src/grpc/_adapter/_c/module.c (hunk 10 above)
Comment (ctiller) [inline src/python/src/grpc/_adapter/_c/module.c]: git autodetects moves... there's no user control over this.

**Comment by soltanmm** [comment]
It was a trashed local repo state.

File: src/python/src/grpc/_adapter/_c/module.c (hunk 10 above)
Comment (soltanmm) [inline src/python/src/grpc/_adapter/_c/module.c]: :-(

**Comment by nathanielmanistaatgoogle** [comment]
@Yhg1s I'd appreciate your eyes on this as much as they were on the first wrapping code back in February - will you be able to take a look or should we chase up @gpshead?

**Comment by soltanmm** [comment]
cc #529, as this makes that issue outdated.

**Comment by Yhg1s** [comment]
On Fri, May 22, 2015 at 2:00 AM, Nathaniel Manista <notifications@github.com
> wrote:

> @Yhg1s <https://github.com/Yhg1s> I'd appreciate your eyes on this as
> much as they were on the first wrapping code back in February - will you be
> able to take a look or should we chase up @gpshead
> <https://github.com/gpshead>?
>
I'll take a look tonight.

-- 
Thomas Wouters <thomas@python.org>

Hi! I'm an email virus! Think twice before sending your email to help me
spread!

**Comment by soltanmm** [comment]
@Yhg1s ping

**Comment by nicolasnoble** [comment]
#1809.

**Comment by soltanmm** [comment]
Updated PR to keep up with protobuf upgrades and the pattern set by #1803...

**Comment by nathanielmanistaatgoogle** [comment]
Merging this; it's waited long enough and we can always do a post-commit review later.
